### PR TITLE
feat: New Module `avm/res/web/site`

### PR DESCRIPTION
**Pull request by krbar**
## Description

New Module `avm/res/web/site`, migrated from CARML.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Pipeline references

| Pipeline |
| - |
| [avm.res.web.site](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.web.site.yml) |

### CODE REVIEW

File: avm/res/web/site/basic-publishing-credentials-policy/main.json [hunk 1]
```diff
@@ -0,0 +1,87 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "metadata": {
+    "_generator": {
+      "name": "bicep",
+      "version": "0.23.1.45101",
```
Comment (AlexanderSehr) [inline avm/res/web/site/basic-publishing-credentials-policy/main.json:7]: Be aware - the current latest version is `0.24.24`

File: avm/res/web/site/tests/e2e/functionApp.defaults/main.test.bicep [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    name: '${namePrefix}${serviceShort}001'
+    location: location
+    kind: 'functionapp'
+    serverFarmResourceId: nestedDependencies.outputs.serverFarmResourceId
+    siteConfig: {
+      alwaysOn: true
+    }
```
Comment (AlexanderSehr) [inline avm/res/web/site/tests/e2e/functionApp.defaults/main.test.bicep:58]: This is just a best practice, right

File: avm/res/web/site/main.bicep [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+  logCategoriesAndGroups: {
+    @description('Optional. Name of a Diagnostic Log category for a resource type this setting is applied to. Set the specific logs to collect here.')
+    category: string?
+
+    @description('Optional. Name of a Diagnostic Log category group for a resource type this setting is applied to. Set to \'AllLogs\' to collect all logs.')
+    categoryGroup: string?
+  }[]?
+
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
```
Comment (AlexanderSehr) [inline avm/res/web/site/main.bicep:545]:
```suggestion
  @description('Optional. The name of metrics that will be streamed. "allMetrics" includes all possible metrics for the resource. Set to \'\' to disable metric collection.')
```

File: avm/res/web/site/main.bicep [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+@description('Optional. The diagnostic settings of the service.')
+param diagnosticSettings diagnosticSettingType
+
+@description('Optional. To enable client certificate authentication (TLS mutual authentication).')
+param clientCertEnabled bool = false
+
+@description('Optional. Client certificate authentication comma-separated exclusion paths.')
+param clientCertExclusionPaths string?
+
+@description('Optional. This composes with ClientCertEnabled setting.</p>- ClientCertEnabled: false means ClientCert is ignored.</p>- ClientCertEnabled: true and ClientCertMode: Required means ClientCert is required.</p>- ClientCertEnabled: true and ClientCertMode: Optional means ClientCert is optional or accepted.')
```
Comment (AlexanderSehr) [inline avm/res/web/site/main.bicep:102]:
```suggestion
@description('''
Optional. This composes with ClientCertEnabled setting.
- ClientCertEnabled=false means ClientCert is ignored.
- ClientCertEnabled=true and ClientCertMode=Required means ClientCert is required.
- ClientCertEnabled=true and ClientCertMode=Optional means ClientCert is optional or accepted.
''')
```

File: avm/res/web/site/main.bicep [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+    hostNameSslStates: slot.?hostNameSslStates
+    hyperV: contains(slot, 'hyperV') ? slot.hyperV : false
+    publicNetworkAccess: contains(slot, 'publicNetworkAccess') ? slot.publicNetworkAccess : ((!empty(slot.?privateEndpoints) || !empty(privateEndpoints)) ? 'Disabled' : 'Enabled')
+    redundancyMode: contains(slot, 'redundancyMode') ? slot.redundancyMode : 'None'
+    vnetContentShareEnabled: contains(slot, 'vnetContentShareEnabled') ? slot.vnetContentShareEnabled : false
+    vnetImagePullEnabled: contains(slot, 'vnetImagePullEnabled') ? slot.vnetImagePullEnabled : false
+    vnetRouteAllEnabled: contains(slot, 'vnetRouteAllEnabled') ? slot.vnetRouteAllEnabled : false
+    hybridConnectionRelays: slot.?hybridConnectionRelays
+  }
+}]
```
Comment (AlexanderSehr) [inline avm/res/web/site/main.bicep:293]:
```suggestion
module app_slots 'slot/main.bicep' = [for (slot, index) in (slots ?? []): {
  name: '${uniqueString(deployment().name, location)}-Slot-${slot.name}'
  params: {
    name: slot.name
    appName: app.name
    location: location
    kind: kind
    serverFarmResourceId: serverFarmResourceId
    httpsOnly: slot.?httpsOnly ?? httpsOnly
    appServiceEnvironmentResourceId: appServiceEnvironmentResourceId
    clientAffinityEnabled: slot.?clientAffinityEnabled ?? clientAffinityEnabled
    managedIdentities: slot.?managedIdentities ?? managedIdentities
    keyVaultAccessIdentityResourceId: slot.?keyVaultAccessIdentityResourceId ?? keyVaultAccessIdentityResourceId
    storageAccountRequired: slot.?storageAccountRequired ?? storageAccountRequired
    virtualNetworkSubnetId: slot.?virtualNetworkSubnetId ?? virtualNetworkSubnetId
    siteConfig: slot.?siteConfig ?? siteConfig
    storageAccountResourceId: slot.?storageAccountResourceId ?? storageAccountResourceId
    appInsightResourceId: slot.?appInsightResourceId ?? appInsightResourceId
    setAzureWebJobsDashboard: slot.?setAzureWebJobsDashboard ?? setAzureWebJobsDashboard
    authSettingV2Configuration: slot.?authSettingV2Configuration ?? authSettingV2Configuration
    diagnosticSettings: slot.?diagnosticSettings
    roleAssignments: slot.?roleAssignments ?? roleAssignments
    appSettingsKeyValuePairs: slot.?appSettingsKeyValuePairs ?? appSettingsKeyValuePairs
    basicPublishingCredentialsPolicies: slot.?basicPublishingCredentialsPolicies ?? basicPublishingCredentialsPolicies
    lock: slot.?lock ?? lock
    privateEndpoints: slot.?privateEndpoints ?? privateEndpoints
    tags: slot.?tags ?? tags
    clientCertEnabled: slot.?clientCertEnabled
    clientCertExclusionPaths: slot.?clientCertExclusionPaths
    clientCertMode: slot.?clientCertMode
    cloningInfo: slot.?cloningInfo
    containerSize: slot.?containerSize
    customDomainVerificationId: slot.?customDomainVerificationId
    dailyMemoryTimeQuota: slot.?dailyMemoryTimeQuota
    enabled: slot.?enabled
    enableTelemetry: slot.?enableTelemetry ?? enableTelemetry
    hostNameSslStates: slot.?hostNameSslStates
    hyperV: slot.?hyperV
    publicNetworkAccess: slot.?publicNetworkAccess ?? ((!empty(slot.?privateEndpoints) || !empty(privateEndpoints)) ? 'Disabled' : 'Enabled')
    redundancyMode: slot.?redundancyMode
    vnetContentShareEnabled: slot.?vnetContentShareEnabled
    vnetImagePullEnabled: slot.?vnetImagePullEnabled
    vnetRouteAllEnabled: slot.?vnetRouteAllEnabled
    hybridConnectionRelays: slot.?hybridConnectionRelays
  }
}]
```

File: avm/res/web/site/slot/main.bicep [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+}]
+
+module slot_privateEndpoints 'br/public:avm/res/network/private-endpoint:0.3.2' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
+  name: '${uniqueString(deployment().name, location)}-Slot-PrivateEndpoint-${index}'
+  params: {
+    privateLinkServiceConnections: [
+      {
+        name: slot.name
+        properties: {
+          privateLinkServiceId: app.id
```
Comment (AlexanderSehr) [inline avm/res/web/site/slot/main.bicep:308]:
```suggestion
          privateLinkServiceId: app.id // Must be set on the WebApp and not the slot
```

File: avm/res/web/site/slot/main.bicep [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+          groupIds: [
+            privateEndpoint.?service ?? 'sites-${slot.name}'
```
Comment (AlexanderSehr) [inline avm/res/web/site/slot/main.bicep:310]:
```suggestion
            privateEndpoint.?service ?? 'sites-${slot.name}' // The required syntax to create the private endpoint for a specific slot
```

File: avm/res/web/site/tests/e2e/webApp.max/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            name: 'scm'
+            allow: false
+          }
+        ]
+        roleAssignments: [
+          {
+            roleDefinitionIdOrName: 'Reader'
+            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
+            principalType: 'ServicePrincipal'
+          }
```
Comment (AlexanderSehr) [inline avm/res/web/site/tests/e2e/webApp.max/main.test.bicep:133]:
```suggestion
          {
            roleDefinitionIdOrName: 'Owner'
            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
            principalType: 'ServicePrincipal'
          }
          {
            roleDefinitionIdOrName: 'b24988ac-6180-42a0-ab88-20f7382dd24c'
            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
            principalType: 'ServicePrincipal'
          }
          {
            roleDefinitionIdOrName: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
            principalId: nestedDependencies.outputs.managedIdentityPrincipalId
            principalType: 'ServicePrincipal'
          }
```

File: avm/res/web/site/tests/e2e/functionApp.defaults/main.test.bicep (hunk 2 above)
Comment (krbar) [reply avm/res/web/site/tests/e2e/functionApp.defaults/main.test.bicep:58]: Correct

File: avm/res/web/site/basic-publishing-credentials-policy/main.json (hunk 1 above)
Comment (krbar) [reply avm/res/web/site/basic-publishing-credentials-policy/main.json]: Thanks., updated

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**